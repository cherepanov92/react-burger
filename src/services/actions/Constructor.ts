import { OrderedIngredient } from '../../utils/types';

export const ADD_INGREDIENT: 'ADD_INGREDIENT' = 'ADD_INGREDIENT';
export const REMOVE_INGREDIENT: 'REMOVE_INGREDIENT' = 'REMOVE_INGREDIENT';
export const MOVE_INGREDIENT: 'MOVE_INGREDIENT' = 'MOVE_INGREDIENT';

interface IAddIngredient {
    readonly type: typeof ADD_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IRemoveIngredient {
    readonly type: typeof REMOVE_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IMoveIngredient {
    readonly type: typeof MOVE_INGREDIENT;
    ingredient: OrderedIngredient;
    newIndex: number;
}

export type TIngredientActions = IAddIngredient | IRemoveIngredient | IMoveIngredient;