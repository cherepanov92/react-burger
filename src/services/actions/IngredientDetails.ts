import { IngredientType } from '../../utils/types';

export const ADD_INGREDIENT_DETAILS: 'ADD_INGREDIENT_DETAILS' = 'ADD_INGREDIENT_DETAILS';
export const REMOVE_INGREDIENT_DETAILS: 'REMOVE_INGREDIENT_DETAILS' = 'REMOVE_INGREDIENT_DETAILS';

interface IAddIngredientDetails {
    readonly type: typeof ADD_INGREDIENT_DETAILS;
    ingredient: IngredientType;
}

interface IRemoveIngredientDetails {
    readonly type: typeof REMOVE_INGREDIENT_DETAILS;
}

export type TIngredientDetailsActions = IAddIngredientDetails | IRemoveIngredientDetails;