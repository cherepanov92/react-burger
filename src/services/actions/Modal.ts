import { EnumModalType } from '../../utils/types';

export const APPEND_MODAL_TYPE: 'APPEND_MODAL_TYPE' = 'APPEND_MODAL_TYPE';
export const APPEND_ERROR_MODAL_TYPE: 'APPEND_ERROR_MODAL_TYPE' = 'APPEND_ERROR_MODAL_TYPE';
export const REMOVE_MODAL_TYPE: 'REMOVE_MODAL_TYPE' = 'REMOVE_MODAL_TYPE';

export interface IAppendModalType {
    readonly type: typeof APPEND_MODAL_TYPE;
    modalType: EnumModalType;
}

interface IAppendErrorModalType {
    readonly type: typeof APPEND_ERROR_MODAL_TYPE;
    status?: string | null;
    message: string | null;
}

interface IRemoveModalTypeE {
    readonly type: typeof REMOVE_MODAL_TYPE;
}

export type TModalActions = IAppendModalType | IRemoveModalTypeE | IAppendErrorModalType;