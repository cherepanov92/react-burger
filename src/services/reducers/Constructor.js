import { ADD_INGREDIENT, MOVE_INGREDIENT, REMOVE_INGREDIENT } from '../actions/Constructor';

const getIngredientWithOrderHash = ingredient => ({
    ...ingredient,
    orderId: ingredient.orderId ? ingredient.orderId : Math.floor(Math.random() * 10000000)
});

export default function ConstructorReducer(state = {}, action) {
    switch (action.type) {
        case ADD_INGREDIENT:
            if (action.ingredient.type === 'bun') {
                const totalPriceWithoutBun = state.bun ? state.totalPrice - state.bun.price * 2 : state.totalPrice;

                return {
                    ...state,
                    bun: action.ingredient,
                    totalPrice: totalPriceWithoutBun + action.ingredient.price * 2
                };
            }

            return {
                ...state,
                ingredients: [
                    ...state.ingredients,
                    { ...getIngredientWithOrderHash(action.ingredient), orderIndex: state.ingredients.length }
                ],
                totalPrice: state.totalPrice + action.ingredient.price
            };
        case REMOVE_INGREDIENT:
            if (!state.ingredients.length) {
                return state;
            }

            return {
                ...state,
                ingredients: [
                    ...state.ingredients.filter(ingredient => ingredient.orderId !== action.ingredient.orderId)
                ],
                totalPrice: state.totalPrice - action.ingredient.price
            };
        case MOVE_INGREDIENT:
            if (state.ingredients[action.newIndex].orderId === action.ingredient?.orderId) {
                return state;
            }

            const OrderListWithoutMovedItem = [...state.ingredients].filter(
                item => item.orderId !== action.ingredient?.orderId
            );

            const newIngredientsOrderList = [
                ...OrderListWithoutMovedItem.slice(0, action.newIndex),
                getIngredientWithOrderHash(action.ingredient),
                ...OrderListWithoutMovedItem.slice(action.newIndex, state.ingredients.length + 1)
            ];

            return {
                ...state,
                ingredients: newIngredientsOrderList.map((item, index) => ({ ...item, orderIndex: index }))
            };

        default:
            return state;
    }
}
