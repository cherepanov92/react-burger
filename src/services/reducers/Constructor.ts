import { ADD_INGREDIENT, MOVE_INGREDIENT, REMOVE_INGREDIENT } from '../actions/Constructor';
import { EnumIngredientType, orderedIngredient } from '../../utils/types';

type constructorState = {
    bun?: orderedIngredient;
    ingredients: orderedIngredient[];
    totalPrice: number;
};

export default function constructorReducer(
    state: constructorState = {
        bun: undefined,
        ingredients: [],
        totalPrice: 0
    },
    action: any
) {
    const getIngredientWithOrderHash = (ingredient: orderedIngredient) => ({
        ...ingredient,
        orderId: ingredient.orderId ? ingredient.orderId : Math.floor(Math.random() * 10000000)
    });

    switch (action.type) {
        case ADD_INGREDIENT:
            if (action.ingredient.type === EnumIngredientType.BUN) {
                const totalPriceWithoutBun = state.bun ? state.totalPrice - state.bun.price * 2 : state.totalPrice;

                return {
                    ...state,
                    bun: action.ingredient,
                    totalPrice: totalPriceWithoutBun + action.ingredient.price * 2
                };
            }

            return {
                ...state,
                ingredients: [
                    ...state.ingredients,
                    { ...getIngredientWithOrderHash(action.ingredient), orderIndex: state.ingredients.length }
                ],
                totalPrice: state.totalPrice + action.ingredient.price
            };
        case REMOVE_INGREDIENT:
            if (!state.ingredients.length) {
                return state;
            }

            return {
                ...state,
                ingredients: [
                    ...state.ingredients.filter(ingredient => ingredient.orderId !== action.ingredient.orderId)
                ],
                totalPrice: state.totalPrice - action.ingredient.price
            };
        case MOVE_INGREDIENT:
            if (state.ingredients[action.newIndex].orderId === action.ingredient?.orderId) {
                return state;
            }

            const OrderListWithoutMovedItem = [...state.ingredients].filter(
                item => item.orderId !== action.ingredient?.orderId
            );

            const newIngredientsOrderList = [
                ...OrderListWithoutMovedItem.slice(0, action.newIndex),
                getIngredientWithOrderHash(action.ingredient),
                ...OrderListWithoutMovedItem.slice(action.newIndex, state.ingredients.length + 1)
            ];

            return {
                ...state,
                ingredients: newIngredientsOrderList.map((item, index) => ({ ...item, orderIndex: index }))
            };

        default:
            return state;
    }
}
