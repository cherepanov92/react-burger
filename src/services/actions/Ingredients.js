import { getIngredients } from "../../utils/api";
import { APPEND_ERROR_MODAL_TYPE } from "./Modal";

export const GET_INGREDIENTS_REQUEST = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED = 'GET_INGREDIENTS_FAILED';

export const getIngredientsData = () => {
    return function(dispatch) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
        .then(res => {
            if (res && res.success) {
                dispatch({
                    type: GET_INGREDIENTS_SUCCESS,
                    items: res.data
                });
            } else {
                dispatch({
                    type: GET_INGREDIENTS_FAILED
                });
            }
        })
        .catch(err => {
            dispatch({
                type: APPEND_ERROR_MODAL_TYPE,
                errorStatus: err.status
            })
        })
    };
}