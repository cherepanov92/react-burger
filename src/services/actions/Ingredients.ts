import { getIngredients } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { IError, IngredientType } from '../../utils/types';
import { Dispatch } from 'react';

export const GET_INGREDIENTS_REQUEST: 'GET_INGREDIENTS_REQUEST' = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS: 'GET_INGREDIENTS_SUCCESS' = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED: 'GET_INGREDIENTS_FAILED' = 'GET_INGREDIENTS_FAILED';

interface IGetIngredientsRequest {
    readonly type: typeof GET_INGREDIENTS_REQUEST;
}

interface IGetIngredientsSuccess {
    readonly type: typeof GET_INGREDIENTS_SUCCESS;
    items: IngredientType[];
}

interface IGetIngredientsFailed {
    readonly type: typeof GET_INGREDIENTS_FAILED;
}

export type TIngredientsActions = IGetIngredientsRequest | IGetIngredientsSuccess | IGetIngredientsFailed;

export const getIngredientsData = () => {
    return function (dispatch: Dispatch<TIngredientsActions | IError>) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_INGREDIENTS_SUCCESS,
                        items: res.data
                    });
                } else {
                    dispatch({
                        type: GET_INGREDIENTS_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
