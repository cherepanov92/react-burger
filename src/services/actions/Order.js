import { sendOrder } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE } from './Modal';

export const SENT_ORDER_REQUEST = 'SENT_ORDER_REQUEST';
export const SENT_ORDER_SUCCESS = 'SENT_ORDER_SUCCESS';
export const SENT_ORDER_FAILED = 'SENT_ORDER_FAILED';

export const sendOrderRequest = orderList => {
    return function (dispatch) {
        dispatch({
            type: SENT_ORDER_REQUEST
        });
        sendOrder(orderList)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: SENT_ORDER_SUCCESS,
                        name: res.name,
                        number: res.order.number
                    });
                    dispatch({ type: APPEND_MODAL_TYPE, modalType: 'order' });
                } else {
                    dispatch({
                        type: SENT_ORDER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
