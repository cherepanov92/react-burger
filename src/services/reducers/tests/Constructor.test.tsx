import React from 'react';
import renderer from 'react-test-renderer';
import constructorReducer, { InitialConstructorState as initialState } from '../Constructor';
import {ADD_INGREDIENT, MOVE_INGREDIENT, REMOVE_INGREDIENT} from '../../actions/Constructor';
import { EnumIngredientType, OrderedIngredient } from '../../../utils/types';
import { getIngredientWithOrderHash } from '../../../utils/helpers';

const getIngredientItem = (
    type: EnumIngredientType = EnumIngredientType.BUN,
    orderId: number = 1,
    orderIndex: number = 1
): OrderedIngredient => ({
    _id: '60d3b41abdacab0026a733c6',
    name: 'noName ingredient',
    type: type,
    proteins: 80,
    fat: 24,
    carbohydrates: 53,
    calories: 420,
    price: 333,
    image: 'https://code.s3.yandex.net/react/code/bun-02.png',
    image_mobile: 'https://code.s3.yandex.net/react/code/bun-02-mobile.png',
    image_large: 'https://code.s3.yandex.net/react/code/bun-02-large.png',
    __v: 0,
    orderIndex: orderIndex,
    orderId: orderId,
    index: 123123
});

const basicBun = getIngredientItem();
const basicSauce = getIngredientItem(EnumIngredientType.SAUCE);
const firstSauce = getIngredientItem(EnumIngredientType.SAUCE, 0, 0);
const secondSauce = getIngredientItem(EnumIngredientType.SAUCE, 1, 1);

const basicOrder = {
    bun: basicBun,
    ingredients: [firstSauce, secondSauce],
    totalPrice: 1332
};

describe('ConstructorReducer:', () => {
    it('Начальное состояние', () => {
        expect(constructorReducer(initialState, {} as any )).toEqual(initialState);
        const tree = renderer.create(<>Начальное состояние</>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Добавление соуса или добавки в закакз', () => {
        expect(
            constructorReducer(initialState, {
                type: ADD_INGREDIENT,
                ingredient: basicSauce
            })
        ).toEqual({
            ...initialState,
            ingredients: [{ ...getIngredientWithOrderHash(basicSauce), orderIndex: initialState.ingredients.length }],
            totalPrice: initialState.totalPrice + basicSauce.price
        });
    });

    it('Добавление булки в закакз', () => {
        expect(
            constructorReducer(initialState, {
                type: ADD_INGREDIENT,
                ingredient: basicBun
            })
        ).toEqual({
            ...initialState,
            bun: basicBun,
            totalPrice: initialState.totalPrice + basicBun.price * 2
        });
    });

    it('Удаление соуса или добавки из закакза', () => {
        expect(
            constructorReducer(basicOrder, {
                type: REMOVE_INGREDIENT,
                ingredient: firstSauce
            })
        ).toEqual({
            ...basicOrder,
            ingredients: [secondSauce],
            totalPrice: basicOrder.totalPrice - firstSauce.price
        });
    });

    it('Перемещение соуса или добавки в закакзе', () => {
        expect(
            constructorReducer(basicOrder, {
                type: MOVE_INGREDIENT,
                ingredient: firstSauce,
                newIndex: 1
            })
        ).toEqual({
            ...basicOrder,
            ingredients: [
                {...secondSauce,orderId: 1, orderIndex:0},
                {...firstSauce,orderId: Math.floor(Date.now() / 1000), orderIndex:1}
            ],
            totalPrice: basicOrder.totalPrice
        });
    });
});
