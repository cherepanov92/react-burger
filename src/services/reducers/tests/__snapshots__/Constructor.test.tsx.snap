// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConstructorReducer: Начальное состояние 1`] = `"Начальное состояние"`;
